embedded-tests/src/test/java/org/apache/druid/testing/embedded/compact/CompactionTestBase.java

```diff
@@ -0,0 +1,134 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.testing.embedded.compact;
+
+import org.apache.druid.indexing.common.task.TaskBuilder;
+import org.apache.druid.indexing.overlord.Segments;
+import org.apache.druid.java.util.common.guava.Comparators;
+import org.apache.druid.testing.embedded.EmbeddedBroker;
+import org.apache.druid.testing.embedded.EmbeddedClusterApis;
+import org.apache.druid.testing.embedded.EmbeddedCoordinator;
+import org.apache.druid.testing.embedded.EmbeddedDruidCluster;
+import org.apache.druid.testing.embedded.EmbeddedHistorical;
+import org.apache.druid.testing.embedded.EmbeddedIndexer;
+import org.apache.druid.testing.embedded.EmbeddedOverlord;
+import org.apache.druid.testing.embedded.EmbeddedRouter;
+import org.apache.druid.testing.embedded.junit5.EmbeddedClusterTestBase;
+import org.apache.druid.timeline.DataSegment;
+import org.joda.time.Interval;
+import org.junit.jupiter.api.Assertions;
+
+import java.io.Closeable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public abstract class CompactionTestBase extends EmbeddedClusterTestBase
+{
+  protected final EmbeddedOverlord overlord = new EmbeddedOverlord();
+  protected final EmbeddedCoordinator coordinator = new EmbeddedCoordinator();
+
+  @Override
+  protected EmbeddedDruidCluster createCluster()
+  {
+    return EmbeddedDruidCluster.withEmbeddedDerbyAndZookeeper()
+                               .useLatchableEmitter()
+                               .addServer(overlord)
+                               .addServer(coordinator)
+                               .addServer(new EmbeddedIndexer())
+                               .addServer(new EmbeddedBroker())
+                               .addServer(new EmbeddedHistorical())
+                               .addServer(new EmbeddedRouter());
+  }
+
+  /**
+   * Deletes all the data for the given datasource so that compaction tasks for
+   * this datasource do not take up task slots unnecessarily.
+   */
+  protected Closeable unloader(String dataSource)
+  {
+    return () -> {
+      overlord.bindings().segmentsMetadataStorage().markAllSegmentsAsUnused(dataSource);
+    };
+  }
+
+  /**
+   * Creates a Task using the given builder and runs it.
+   *
+   * @return ID of the task.
+   */
+  protected String runTask(TaskBuilder<?, ?, ?> taskBuilder, String dataSource)
```

Comment:
IMO, `protected` methods on a base class aren't the best way to do utility APIs. Other test cases might want some of these utility methods, but not to be "compaction tests". Also, some tests might want to extend multiple base classes, and this approach makes it impossible.

With MSQ tests we have `EmbeddedMSQApis`, something that collects together utility APIs without using a base class approach. Something similar might work here?